File: sdk/rust/cyt-indexer/src/catalog_builder.rs
use crate::build::{CatalogIndex, build_catalog_index};
use crate::catalog_io::write_catalog_index;
use serde_json::Value;
use std::collections::HashMap;
use std::path::PathBuf;

pub struct CatalogBuilder {
    pub tools: Vec<Value>,
    pub all_enums: Vec<Value>,
    index: Option<CatalogIndex>,
    lookup: HashMap<(String, String), usize>,
    pub memory_only: bool,
    pub output_dir: Option<PathBuf>,
}

impl CatalogBuilder {
    #[must_use]
    pub fn new_with_options(memory_only: Option<bool>, output_dir: Option<PathBuf>) -> Self {
        let cfg = crate::paths::snapshot();
        Self {
            tools: Vec::new(),
            all_enums: Vec::new(),
            index: None,
            lookup: HashMap::new(),
            memory_only: memory_only.unwrap_or(cfg.builder_memory_only),
            output_dir,
        }
    }

    #[must_use]
    pub fn new(memory_only: bool, output_dir: Option<PathBuf>) -> Self {
        Self::new_with_options(Some(memory_only), output_dir)
    }

    pub fn add_tool(&mut self, entry: Value) {
        let server = entry
            .get("server")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();
        let tool = entry
            .get("tool")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();
        let idx = self.tools.len();
        self.lookup.insert((server, tool), idx);
        if let Some(schema) = entry
            .get("full_schema")
            .and_then(|fs| fs.get("inputSchema"))
        {
            self.all_enums.extend(crate::paths::collect_enums(schema));
        }
        self.tools.push(entry);
        self.index = None;
    }

    #[must_use]
    pub fn get_tool_info(&self, server_name: &str, tool_name: &str) -> Option<&Value> {
        let key = (server_name.to_string(), tool_name.to_string());
        self.lookup.get(&key).and_then(|&i| self.tools.get(i))
    }

    #[must_use]
    pub fn build_index(&mut self) -> &CatalogIndex {
        self.index
            .get_or_insert_with(|| build_catalog_index(&self.tools, &self.all_enums))
    }

    /// Persist the catalog index to disk when not in `memory_only` mode.
    ///
    /// # Errors
    /// Returns an error string when catalog files cannot be written to `output_dir`.
    pub fn write_catalog(&mut self) -> Result<&CatalogIndex, String> {
        let memory_only = self.memory_only;
        let output_dir = self.output_dir.clone();
        let index_ref = self.build_index();
        if !memory_only {
            let default_dir = crate::paths::default_catalog_dir();
            let dir = output_dir.as_deref().unwrap_or(&default_dir);
            write_catalog_index(index_ref, dir, crate::paths::write_catalog_prune())?;
        }
        Ok(index_ref)
    }

    pub fn to_catalog_dict(&mut self) -> Value {
        self.build_index().to_catalog_dict()
    }

    pub fn to_catalog_dict_with_prefix(&mut self, catalog_prefix: &str) -> Value {
        self.build_index()
            .to_catalog_dict_with_prefix(catalog_prefix)
    }
}

impl Default for CatalogBuilder {
    fn default() -> Self {
        Self::new_with_options(None, None)
    }
}

File: sdk/go/cytindexer/catalog_builder.go
package cytindexer

// CatalogBuilder incrementally builds a decomposed catalog index.
type CatalogBuilder struct {
	Tools       []any
	AllEnums    []any
	index       *CatalogIndex
	lookup      map[[2]string]int
	MemoryOnly  bool
	OutputDir   string
}

// NewCatalogBuilder creates a builder with explicit memory-only and output dir settings.
func NewCatalogBuilder(memoryOnly bool, outputDir string) *CatalogBuilder {
	return NewCatalogBuilderWithOptions(&memoryOnly, outputDir)
}

// NewCatalogBuilderWithOptions creates a builder using optional memory-only override.
func NewCatalogBuilderWithOptions(memoryOnly *bool, outputDir string) *CatalogBuilder {
	cfg := PathSnapshot()
	mo := cfg.BuilderMemoryOnly
	if memoryOnly != nil {
		mo = *memoryOnly
	}
	return &CatalogBuilder{
		lookup:     make(map[[2]string]int),
		MemoryOnly: mo,
		OutputDir:  outputDir,
	}
}

// AddTool adds a catalog tool entry.
func (b *CatalogBuilder) AddTool(entry any) {
	obj, _ := AsObject(entry)
	server := StrField(obj, "server")
	tool := StrField(obj, "tool")
	idx := len(b.Tools)
	b.lookup[[2]string{server, tool}] = idx
	if fs, ok := AsObject(obj["full_schema"]); ok {
		if schema, ok := fs["inputSchema"]; ok {
			b.AllEnums = append(b.AllEnums, CollectEnums(schema)...)
		} else if schema, ok := fs["input_schema"]; ok {
			b.AllEnums = append(b.AllEnums, CollectEnums(schema)...)
		}
	}
	b.Tools = append(b.Tools, entry)
	b.index = nil
}

// GetToolInfo returns a tool entry by server and tool name.
func (b *CatalogBuilder) GetToolInfo(serverName, toolName string) (any, bool) {
	idx, ok := b.lookup[[2]string{serverName, toolName}]
	if !ok || idx >= len(b.Tools) {
		return nil, false
	}
	return b.Tools[idx], true
}

// BuildIndex builds and caches the catalog index.
func (b *CatalogBuilder) BuildIndex() *CatalogIndex {
	if b.index == nil {
		idx := BuildCatalogIndex(b.Tools, b.AllEnums)
		b.index = &idx
	}
	return b.index
}

// WriteCatalog builds and optionally writes the catalog to disk.
func (b *CatalogBuilder) WriteCatalog() (*CatalogIndex, error) {
	index := b.BuildIndex()
	if !b.MemoryOnly {
		dir := b.OutputDir
		if dir == "" {
			dir = DefaultCatalogDir()
		}
		if err := WriteCatalogIndex(index, dir, WriteCatalogPrune()); err != nil {
			return nil, err
		}
	}
	return b.index, nil
}

// ToCatalogDict returns the catalog as a dictionary.
func (b *CatalogBuilder) ToCatalogDict() map[string]any {
	return b.BuildIndex().ToCatalogDict()
}

// ToCatalogDictWithPrefix returns the catalog dict with a custom prefix.
func (b *CatalogBuilder) ToCatalogDictWithPrefix(catalogPrefix string) map[string]any {
	return b.BuildIndex().ToCatalogDictWithPrefix(catalogPrefix)
}
